ext/curl/share.c

```diff
@@ -134,6 +136,112 @@ PHP_FUNCTION(curl_share_strerror)
 }
 /* }}} */
 
+/**
+ * Initialize a persistent curl share handle with the given share options, reusing an existing one if found.
+ *
+ * Throws an exception if the share options are invalid.
+ */
+PHP_FUNCTION(curl_persistent_share_init)
+{
+	zval *share_opts = NULL, *entry = NULL;
+	zend_ulong persistent_id = 0;
+
+	php_curlsh *sh;
+
+	CURLSHcode error;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_ARRAY_EX(share_opts, 0, 1)
+	ZEND_PARSE_PARAMETERS_END();
+
+	object_init_ex(return_value, curl_persistent_share_ce);
+	sh = Z_CURL_SHARE_P(return_value);
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(share_opts), entry) {
+		ZVAL_DEREF(entry);
+
+		zend_ulong option = zval_get_long_ex(entry, true);
+
+		if (option == CURL_LOCK_DATA_COOKIE) {
+			zend_throw_exception_ex(
+				NULL,
+				0,
+				"CURL_LOCK_DATA_COOKIE is not allowed with persistent curl share handles"
+			);
+
+			goto error;
+		}
```

Comment:
Might make more sense to use a strict allow-list. This would also make the ID generation easier to verify:

```
switch (option) {
case CURL_LOCK_DATA_COOKIE:
  zend_throw_exception_ex(NULL, 0, "CURL_LOCK_DATA_COOKIE is not allowed with persistent curl share handles");
  goto error;
case CURL_LOCK_DATA_DNS:
  persistent_id |= 1;
  break;
case CURL_LOCK_DATA_CONNECT:
  persistent_id |= 2;
  break;
/* more options here */
default:
  zend_throw_exception_ex(NULL, 0, "Unknown option yada yada");
  goto error;
}
```